### PR TITLE
Add docs and release notes for Build Features API

### PR DESCRIPTION
**Pull request by alllex**
Follow up for
- https://github.com/gradle/gradle/pull/26325

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by alllex** [comment]
@bot-gradle test this

**Comment by alllex** [comment]
@bot-gradle test this

File: subprojects/docs/src/docs/release/notes.md [hunk 1]
```diff
@@ -121,6 +121,17 @@ versionCatalogs.named("libs").findLibrary("assertj-core").ifPresent { assertjCor
...
 
 
+<a name="build-features"></a>
+### Build Features API
+
+The new incubating Build Features API allows plugins to access the status of build features during build time.
+It is designed to help plugin authors and build engineers to adopt Gradle features incrementally.
+
+In the initial release, the API provides information about the Configuration Cache and Isolated Projects features.
+For each build feature it is possible to determine whether it was requested by the user, and if it is active in the current build.
```
Comment (lkasso) [inline subprojects/docs/src/docs/release/notes.md:131]:
```suggestion
For each build feature, it is possible to determine whether the user requested it and if it is active in the current build.
```

**Review by lkasso — APPROVED** [review]
Left a few suggestions.

File: subprojects/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -121,6 +121,17 @@ versionCatalogs.named("libs").findLibrary("assertj-core").ifPresent { assertjCor
...
+
```
[the remark this answers is not in the document]
Comment (alllex) [reply subprojects/docs/src/docs/release/notes.md:132]: I was not sure whether we should include a release notes section at all for this feature. In the current state, it is mainly useful for large ecosystem plugins such as Android Gradle Plugin and Kotlin Gradle Plugin, but not really for others.

In other words, this is intended for very advanced users, and I would not draw more attention to it with actionable-looking snippets. Moreover, the feature is not intended to be used from the DSL (aka build script), but is intended to be used from plugin implementations that would be quite wordy for a release notes section.

In other words, I would try to keep this short, and not add a snippet.

File: subprojects/docs/src/docs/release/notes.md [hunk 3]
```diff
@@ -121,6 +121,17 @@ versionCatalogs.named("libs").findLibrary("assertj-core").ifPresent { assertjCor
 Check the [version catalog API](javadoc/org/gradle/api/artifacts/VersionCatalog.html) for all supported methods.
 
 
+<a name="build-features"></a>
+### Build Features API
+
+The new incubating Build Features API allows plugins to access the status of build features during build time.
+It is designed to help plugin authors and build engineers to adopt Gradle features incrementally.
+
+In the initial release, the API provides information about the Configuration Cache and Isolated Projects features.
```
[the remark this answers is not in the document]
Comment (alllex) [reply subprojects/docs/src/docs/release/notes.md:130]: I intentionally did not provide links here. I think for folks who do read Gradle's release notes, the Configuration Cache is a well known thing. But for the Isolated Projects, we intentionally don't have a dedicated page in our userguide yet.

So, I would keep those just as mentions.

Just as a note, the suggestion you provided uses Asciidoc link syntax, while the release notes is a Markdown file, and the link syntax there is `[]()`. We should have gone with only one of the formats, but there is never time to do a migration.

**Comment by alllex** [comment]
@lkasso, thanks for the suggestions. I applied and resolved almost all. For the ones I didn't apply, I left a comment.

File: subprojects/docs/src/docs/release/notes.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (lkasso) [reply subprojects/docs/src/docs/release/notes.md:130]: > I intentionally did not provide links here.

Sounds good. Let's keep it link free.

> We should have gone with only one of the formats, but there is never time to do a migration.

Thank you for the tip on link syntax, I can enforce Markdown moving forward (I kept seeing both, so I assumed we were ok with both).

**Comment by alllex** [comment]
@bot-gradle test DB

**Comment by alllex** [comment]
@bot-gradle test BD

File: subprojects/docs/src/docs/userguide/optimizing-performance/configuration_cache.adoc [hunk 4]
```diff
@@ -581,6 +581,18 @@ Let your team know they can opt-in by, for example, enabling the configuration c
...
+
+There are additional ways to streamline the migration process for the configuration cache compatibility.
+
+Build logic or plugin implementations can detect if the configuration cache is enabled for a given build, and react to it accordingly.
+The link:{javadocPath}/org/gradle/api/configuration/BuildFeature.html#getActive--[active] status of the configuration cache is provided in the corresponding link:{javadocPath}/org/gradle/api/configuration/BuildFeatures.html#getConfigurationCache--[build feature].
+You can access it by <<custom_gradle_types#service_injection, injecting>> the link:{javadocPath}/org/gradle/api/configuration/BuildFeatures.html[`BuildFeatures`] service into your code.
+
+You can use this information to configure features of your plugin differently or to disable an optional feature that is not yet compatible.
+Another example involves providing additional guidance for your users, should they need to adjust their setup or be informed of temporary limitations.
+
```
Comment (abstratt) [inline subprojects/docs/src/docs/userguide/optimizing-performance/configuration_cache.adoc:595]: 💯

**Review by abstratt — APPROVED** [review]
LGTM, thanks for including the section in the CC docs pointing back at the Build Features doc

File: subprojects/docs/src/docs/userguide/optimizing-performance/configuration_cache.adoc [hunk 5]
```diff
@@ -581,6 +581,18 @@ Let your team know they can opt-in by, for example, enabling the configuration c
 Later on, when more workflows are working, you can flip this around.
 Enable the configuration cache by default, configure CI to disable it, and if required communicate the unsupported workflow(s) for which the configuration cache needs to be disabled.
 
+[[config_cache:adoption:reacting]]
+=== Reacting to the configuration cache in the build
+
+There are additional ways to streamline the migration process for the configuration cache compatibility.
```
Comment (mlopatkin) [inline subprojects/docs/src/docs/userguide/optimizing-performance/configuration_cache.adoc:587]: :thought_balloon: I'm not sure this introductory statement fits here, we're telling about a particular way of streamlining. I'd suggest to remove this paragraph.

**Review by mlopatkin — APPROVED** [review]
LGTM, with just one nitpick

File: subprojects/core-api/src/main/java/org/gradle/api/configuration/BuildFeature.java [hunk 6]
```diff
@@ -25,6 +25,9 @@
  * <p>
  * It is possible to check if the feature is {@link #getActive() active} in the current build.
  * The {@link #getRequested() requested} property shows whether the user opted in or opted out from the feature.
+ * <p>
+ * An instance of this type can be injected into a task, plugin or other object by annotating
```
Comment (bamboo) [inline subprojects/core-api/src/main/java/org/gradle/api/configuration/BuildFeature.java:29]: 👍

File: subprojects/docs/src/snippets/developingPlugins/reactingToBuildFeatures/groovy/buildSrc/src/main/java/MyPlugin.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.configuration.BuildFeatures;
+
+import javax.inject.Inject;
+
+// tag::my-plugin[]
+public abstract class MyPlugin implements Plugin<Project> {
+
+    private final MyReport myReport = new MyReport();
```
Comment (bamboo) [inline subprojects/docs/src/snippets/developingPlugins/reactingToBuildFeatures/groovy/buildSrc/src/main/java/MyPlugin.java:10]: ☁️ I'd remove this field and move the report instance to the `apply` method so the `BuildFeatures` injection is the first and only thing the user sees at this level

**Review by bamboo — APPROVED** [review]
Left a few suggestions

File: subprojects/docs/src/snippets/developingPlugins/reactingToBuildFeatures/groovy/buildSrc/src/main/java/MyPlugin.java (hunk 7 above)
Comment (alllex) [reply subprojects/docs/src/snippets/developingPlugins/reactingToBuildFeatures/groovy/buildSrc/src/main/java/MyPlugin.java:10]: Great suggestion!

**Comment by alllex** [comment]
@bot-gradle test and merge
